liquidctl/driver/aquacomputer.py

```diff
@@ -0,0 +1,262 @@
+These groups provide RPM speed, voltage, current and power readings. The
+pump additionally exposes +5V and +12V voltage rail readings.
+
+Driver
+------
+Linux has the aquacomputer_d5next driver available since v5.15. Subsequent
+releases have more functionality and support a wider range of devices. If
+present, it's used instead of reading the status reports directly.
+
+Copyright (C) 2022 - Aleksa Savic
+
+SPDX-License-Identifier: GPL-3.0-or-later
+"""
+
+# uses the psf/black style
+
+import logging
+
+from liquidctl.driver.usb import UsbHidDriver
+from liquidctl.error import NotSupportedByDriver
+
+_LOGGER = logging.getLogger(__name__)
+
+_AQC_TEMP_SENSOR_DISCONNECTED = 0x7FFF
+_AQC_FAN_VOLTAGE_OFFSET = 0x02
+_AQC_FAN_CURRENT_OFFSET = 0x04
+_AQC_FAN_POWER_OFFSET = 0x06
+_AQC_FAN_SPEED_OFFSET = 0x08
+
+_AQC_STATUS_READ_ENDPOINT = 0x01
+
+
+def get_unaligned_be16(msg, offset):
+    return msg[offset] << 8 | msg[offset + 1]
+
+
+class Aquacomputer(UsbHidDriver):
+    # Support for hwmon: aquacomputer_d5next, sensors - 5.15+
+
+    DEVICE_D5NEXT = "D5 Next"
+
+    DEVICE_INFO = {
+        DEVICE_D5NEXT: {
+            "type": DEVICE_D5NEXT,
+            "fan_sensors": [0x6C, 0x5F],
+            "temp_sensors": [0x57],
+            "plus_5v_voltage": 0x39,
+            "plus_12v_voltage": 0x37,
+            "temp_sensors_label": ["Liquid temperature"],
+            "fan_speed_label": ["Pump speed", "Fan speed"],
+            "fan_power_label": ["Pump power", "Fan power"],
+            "fan_voltage_label": ["Pump voltage", "Fan voltage"],
+            "fan_current_label": ["Pump current", "Fan current"],
+            "status_report_length": 0x9E,
+        }
+    }
+
+    SUPPORTED_DEVICES = [
+        (0x0C70, 0xF00E, None, "Aquacomputer D5 Next", {"device_info": DEVICE_INFO[DEVICE_D5NEXT]}),
+    ]
+
+    def __init__(self, device, description, device_info, **kwargs):
+        super().__init__(device, description)
+
+        # Read when necessary
+        self._firmware_version = None
+
+        self._device_info = device_info
+
+    def initialize(self, **kwargs):
+        """Initialize the device and the driver.
+
+        This method should be called every time the system boots, resumes from
+        a suspended state, or if the device has just been (re)connected.  In
+        those scenarios, no other method, except `connect()` or `disconnect()`,
+        should be called until the device and driver has been (re-)initialized.
+
+        Returns None or a list of `(property, value, unit)` tuples, similarly
+        to `get_status()`.
+        """
+
+        fw = self.firmware_version
+        _LOGGER.debug("raw firmware version: %r", fw)
+
+        return [("Firmware version", fw, "")]
+
+    def _get_status_directly(self):
+        msg = self._read()
+
+        sensor_readings = []
+
+        # Read temp sensor values
+        for idx, temp_sensor_offset in enumerate(self._device_info["temp_sensors"]):
```

Comment:
This is probably just as easy to read as is, if not more, but a reminder that you could also do something like:

```suggestion
        for label, offset in zip(self._device_info["temp_sensors_label"], self._device_info["temp_sensors"]):
```

...here and in the similar loops bellow.

Reply:
Thanks, I've been aware of zip() before, but I never had an opportunity to use it (in Python or elsewhere). I converted that loop to use it.

Reply:
Let me start by saying sorry for the noise!

I think it's (more) important to be consistent throughout the file.  And while it's possible to use zip in the other loops too, it'll probably look a bit unwieldy when more than one label is needed per iteration.  Honestly, even in this loop, I think I prefer the look of your original version, thanks to the shorter and more uniform lines.

I also thought that there could be a performance penalty in doing the extra indexed (and checked) access, but I can't actually see that in a quick microbenchmark, at least not with Python 3.10.

So I might revert this back before merging the PR, unless you want to try using zip on the other loops.

Again, sorry!

Reply:
No problem, I'll revert it now to make it easier.

Reply:
Just saw the edit - off the top of my head there will be 8-10 max register offsets (for Octo), so we're good.